subsys/bluetooth/services/ots/ots_client.c

```diff
@@ -1156,7 +1196,70 @@ static int oacp_read(struct bt_conn *conn,
 
 	return err;
 }
+static int oacp_write(struct bt_conn *conn,
+		     struct bt_otc_internal_instance_t *inst,
+			 uint8_t *buf, size_t len, off_t offset, uint8_t mode)
+{
+	int err;
+	struct bt_gatt_ots_l2cap *l2cap;
+
+	LOG_DBG("");
+	if (!inst->otc_inst->oacp_handle) {
+		BT_DBG("Handle not set");
+		return -EINVAL;
+	} else if (inst->busy) {
+		return -EBUSY;
+	} else if (cur_inst) {
+		return -EBUSY;
+	}
+
+	/* TODO: How do we ensure that the L2CAP is connected in time for the
+	 * transfer?
+	 */
+
+	err = bt_gatt_ots_l2cap_connect(conn, &l2cap);
+	if (err) {
+		BT_DBG("Could not connect l2cap: %d", err);
+		return err;
+	}
+	l2cap->tx_done = write_obj_tx_done;
+	l2cap->rx_done = rx_done;
+	l2cap->closed  = chan_closed;
+	l2cap->tx.data = buf;
+	l2cap->tx.len = len;
+
+	net_buf_simple_reset(&otc_tx_buf);
+
+	/* OP Code */
+	net_buf_simple_add_u8(&otc_tx_buf, BT_GATT_OTS_OACP_PROC_WRITE);
+
+	/* Offset */
+	net_buf_simple_add_le32(&otc_tx_buf, offset);
```

Comment:
`net_buf_simple_add_le32` takes a `uint32_t` and `offset` is a `off_t`, so we will need some value validation before adding it with `net_buf_simple_add_le32`.

Reply:
/* OTS_v10.pdf Table 3.9: Object Action Control Point Procedure Requirements
	 *	Offset and Length field are UINT32 Length 
	 */

	/* Offset */
	offset = offset & UINT32_MAX;
	net_buf_simple_add_le32(&otc_tx_buf, offset);

	/* Len */
	len = len & UINT32_MAX;
	net_buf_simple_add_le32(&otc_tx_buf, len);

Reply:
Something like that, yes, but keep in mind that `off_t` is a signed value.

Reply:
Actually, I guess the right solution would be to return an error instead of silently truncating the values here, as this is triggered from an application API, right?

Reply:
Check before add to net buf.
	if (offset < UINT32_MAX) {
		net_buf_simple_add_le32(&otc_tx_buf, offset);
	} else {
		BT_ERR("offset exceed UINT32");
		return -EINVAL;
	}

	/* Len */
	if (len < UINT32_MAX) {
		net_buf_simple_add_le32(&otc_tx_buf, len);
	} else {
		BT_ERR("offset exceed UINT32");
		return -EINVAL;
	}
But would it be easier to use uint32_t as type of len and offset?

Reply:
It would be easier, yes. See https://github.com/zephyrproject-rtos/zephyr/pull/45927#discussion_r887681112

Reply:
It is fine that we now verify the value of the offset, but is there any reason to use `off_t` for this internal function instead of `uint32_t`, which better matches the OTS specification? 

I think `off_t` might be better for the public API, but internally we can move it to a type that's easier to work with, especially since we are using it as a `uint32_t` when calling `net_buf_simple_add_le32`.

Reply:
As discussion, use size_t and off_t for public API, use uint32_t for internal function.